### DISCUSSION TITLE
WebGL generateRenderInstructions performance

### DISCUSSION THREAD

**Discussion by guldbran** [openlayers/openlayers#15672] [Q&A]
Hi,
I'm currently trying to wrap my head around the WebGL renderer, and stumbled upon something I don't quite understand.

Looking at the WebGL Vector Layers-example, there seems to be two major performance hotspots.
![image](https://github.com/openlayers/openlayers/assets/10121077/c7e1afb2-02cb-4673-9050-f2f98275d1f9)

1. rendertarget.readAll(). MDN mentions readPixels() as potentially blocking if not correctly: (https://developer.mozilla.org/en-US/docs/Web/API/WebGL_API/WebGL_best_practices). The OL implementation looks okay though, so maybe there is little to be done here.

2. renderinstructions.generatePolygon/LinestringRenderInstructions(). The issue here seems to be the transform2D per vertex. I could be wrong, but could it be worth considering moving these transformations to the GPU?

//Kjetil

**Comment by guldbran** [comment]
Bump @jahow 😇

**Reply by jahow** [reply]
Hi! Sorry I was sure to have answered you!

First, thank you for your interest in the renderer and your detailed investigation. Transforming coordinates is indeed done on the CPU, simply because there's currently nothing in the library to do it on the GPU.

A big constraint here is that we don't make any assumption on the coordinate system of the geometries and the view projection, so really both systems could be anything. Reimplementing the transformation logic on the GPU would be quite a significant effort IMO.

On the other hand I'm wondering why the render instructions have to be transformed; they could stay in the original geometry coordinate system, and the transformation could happen on the worker when generating the webgl buffers. That could be a lead worth investigating!
